### PR TITLE
fix: honor per-request seed/offset in sampling kernels

### LINKED ISSUE #1104: FlashInfer 0.2.3+ does not support per-request generators
I'm using FlashInfer 2.5 attention backend with vLLM 0.9.0.1 on CUDA 12.8 and with `VLLM_USE_FLASHINFER_SAMPLER=1`, when temperature deviates between requests, this is what happens:

```
...
1|vllm-server-7011  | (VllmWorker rank=2 pid=16417) WARNING 05-31 03:56:50 [topk_topp_sampler.py:99] FlashInfer 0.2.3+ does not support per-request generators. Falling back to PyTorch-native implementation.
1|vllm-server-7011  | (VllmWorker rank=3 pid=16418) WARNING 05-31 03:56:50 [topk_topp_sampler.py:99] FlashInfer 0.2.3+ does not support per-request generators. Falling back to PyTorch-native implementation.
1|vllm-server-7011  | (VllmWorker rank=7 pid=16422) WARNING 05-31 03:56:50 [topk_topp_sampler.py:99] FlashInfer 0.2.3+ does not support per-request generators. Falling back to PyTorch-native implementation.
1|vllm-server-7011  | (VllmWorker rank=6 pid=16421) WARNING 05-31 03:56:50 [topk_topp_sampler.py:99] FlashInfer 0.2.3+ does not support per-request generators. Falling back to PyTorch-native implementation.
1|vllm-server-7011  | (VllmWorker rank=5 pid=16420) WARNING 05-31 03:56:50 [topk_topp_sampler.py:99] FlashInfer 0.2.3+ does not support per-request generators. Falling back to PyTorch-native implementation.
...
```

I am able to set it to `VLLM_USE_FLASHINFER_SAMPLER=1` and it falls back to Flash Attention sampler.

Please let me know if this is a WIP, thanks!

Cross post:
https://github.com/vllm-project/vllm/issues/18811

### PR DESCRIPTION
**Pull request by zhaoyangwang-nvidia**
## 📌 Description

The sampling kernels accept `seed`/`offset` as per-request tensors, and the Python layer validates them as length 1 or `batch_size`, but every kernel read only `seed_arr[0]` / `offset_arr[0]` and applied that single value to the entire batch. Rows were separated solely by the Philox subsequence (`blockIdx.x`).

So rows did draw from distinct streams, but no row except the first used the seed its caller asked for. A user-specified per-request seed was unreproducible whenever `batch_size > 1`, and a request's tokens depended on where it happened to land in the batch. Concretely, seeds `[1111, 2222, 3333, 4444]` produced the same output as `[1111, 1111, 1111, 1111]`.

This PR indexes the arrays by `blockIdx.x`, matching the convention already used by the other per-request arrays in the same kernels (e.g. `top_k_arr[bx]`). Seven kernels are affected: `SamplingFromLogitsKernel`, `SamplingFromProbKernel`, `TopKSamplingFromProbKernel`, `TopPSamplingFromProbKernel`, `MinPSamplingFromProbKernel`, `TopKTopPSamplingFromProbKernel` and `ChainSpeculativeSampling`.

A length-1 tensor still broadcasts to every row, so existing callers that pass the tensor spelling of a scalar seed are unaffected. The broadcast is resolved in-kernel via a flag rather than by materializing an expanded tensor, since expanding would allocate on every call and break CUDA graph capture — which is precisely the case per-request seeds are needed for.

The length that makes the indexing safe was previously only checked in Python, so a direct FFI caller could read out of bounds. `validate_seed_offset_tensors` now checks it too (and derives the broadcast flag), and seed/offset tensors of differing lengths are rejected in both layers.

**Scope:** this is deliberately limited to making per-request seed/offset take effect. It does not include the in-kernel atomic offset update / round-based offset tracking from #2345; those are separable and can follow if wanted.

## 🔍 Related Issues

- Closes #1104
- Supersedes #2345 — opened as a new PR at @yzh119's suggestion, since he does not currently have bandwidth to carry that one forward. Credit for the original design and the C++/Python plumbing goes to him; this PR reuses that surface and fixes the kernel-side indexing. Downstream, TensorRT-LLM already plumbs per-row Philox `(seed, offset)` end-to-end and is blocked on this behavior.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

Added to `tests/utils/test_sampling.py`, covering all 6 sampling entry points × 2 vocab sizes plus chain-speculative sampling:

- `test_per_request_seed_selects_that_rows_stream` / `test_per_request_offset_selects_that_rows_stream` — running with a mixed seed (offset) tensor must equal, row for row, running each half separately with the corresponding scalar value.
- `test_length_one_seed_tensor_matches_scalar` — a length-1 tensor must broadcast and match the scalar path exactly.
- `test_per_request_seed_reproducibility` — per-row tensors are reproducible across calls.
- `test_seed_offset_tensor_length_validation` — bad lengths and mismatched seed/offset lengths raise instead of reading out of bounds.
- `test_chain_speculative_sampling_per_request_seed` — same guarantee for the rejection-sampling kernel.

Note the kernels intentionally vary the Philox *subsequence* by row, so two rows sharing a seed still differ; the tests assert the property that actually matters — that a row's sample follows its own seed entry — rather than asserting rows with equal seeds are identical.

Verified on B200 (CUDA 13, PyTorch 2.12):

- With the fix: **84 passed**.
- Reverting just the per-row indexing to `idx = 0` to emulate the old behavior: **25 failed**, covering all seven kernels — confirming the tests actually catch the bug rather than passing vacuously.
- Full `tests/utils/test_sampling.py`: **1860 passed, 120 skipped**. Six `test_top_k_top_p_sampling_from_probs_logits_alignment[...128256...]` failures also reproduce on an unmodified `main` at `5a8e62ac`, so they are pre-existing and unrelated to this change.

## Reviewer Notes

The kernel diff is mechanical and identical across all seven kernels — the three-line seed/offset resolution is now a shared `ResolvePhiloxSeedOffset` helper, so the per-kernel change is one call.

Worth a look: `bx` is used as the seed index rather than `row_idx`. Seed/offset are validated against `batch_size` (== `indices.size(0)` when `indices` is given), so `bx` is the batch row and is the correct index, consistent with `top_k_arr[bx]` in the same kernels; `row_idx` is the row of the probs/logits tensor and would be wrong here.

AI-assisted with Claude Code; I've reviewed the change and can walk through the rationale.

## Summary by CodeRabbit

* **New Features**
  * Sampling now supports per-request random seeds and offsets.
  * Length-1 seed and offset tensors are broadcast across the batch.
  * Batch-sized seed and offset tensors allow independent reproducible sampling for each request.
* **Bug Fixes**
  * Invalid seed/offset lengths and mismatched tensor lengths now raise clear validation errors.
* **Tests**
  * Added coverage for broadcasting, per-request reproducibility, validation errors, and speculative sampling.

### CODE REVIEW

**Comment by zhaoyangwang-nvidia** [comment]
@flashinfer-bot run

**Comment by zhaoyangwang-nvidia** [comment]
Hi @yzh119 @kahyunnam @bkryu — could I ask one of you to take a look?

Context: this is the new PR replacing #2345, opened at @yzh119's suggestion since he doesn't have bandwidth to carry that one forward. It's scoped down to the part that fixes the actual bug — indexing `seed_arr`/`offset_arr` by `blockIdx.x` in the 7 sampling kernels, so a per-request seed takes effect. It does not include the atomic offset update / round-based tracking from #2345; those are separable and can follow.

Downstream, TensorRT-LLM already plumbs per-row Philox `(seed, offset)` end-to-end, so today seeds `[1111, 2222, 3333, 4444]` behave identically to `[1111, 1111, 1111, 1111]` — this blocks us from enabling per-request seeded sampling on the FlashInfer-backed path.

Two things worth a reviewer's eye:
- **`bx` vs `row_idx` as the seed index.** seed/offset are validated against `batch_size` (== `indices.size(0)` when `indices` is passed), so `bx` is the batch row and matches `top_k_arr[bx]` in the same kernels. `row_idx` is the probs/logits row and would be the wrong index here.
- **Why the length-1 broadcast is resolved in-kernel** rather than by expanding in Python: `.expand()` yields a non-contiguous view that `CHECK_INPUT` rejects, and `.repeat()` allocates on every call, which breaks CUDA graph capture — the very case per-request seeds are needed for.

I also added a C++-side length check; it was previously only validated in Python, so a direct FFI caller could read out of bounds.

On testing: I verified on B200 that the new tests fail (25 failures across all 7 kernels) when the per-row indexing is reverted to `idx = 0`, so they genuinely catch the bug rather than passing vacuously. The 6 pre-existing `test_top_k_top_p_sampling_from_probs_logits_alignment[...128256...]` failures also reproduce on an unmodified `main` at `5a8e62ac` and are unrelated to this change.

Also — I'm not authorized to start CI (`@flashinfer-bot run` was rejected above), so I'd appreciate it if someone with permissions could kick it off. I've only been able to test on B200 locally, so the wider architecture matrix would be valuable here given this touches shared sampling kernels.

Thanks!

**Comment by yzh119** [comment]
@flashinfer-bot run

**Comment by zhaoyangwang-nvidia** [comment]
Hi @aleozlx @bkryu @cyx-6 @jimmyzho could you help to review this PR? CI is passed.

**Comment by zhaoyangwang-nvidia** [comment]
Hi @kahyunnam @saltyminty @sricketts could you help to review this PR? thanks a lot~

**Comment by aleozlx** [comment]
## PR Review Screening

**CI verdict:** ✅ auto-run ok
**Review category:** `human` (rule fired: C3.2 — shared sampling header and seven kernels, with C3.3 strong)
**Blocking checks:** none
**Release blocker:** 🚨 candidate — a user-supplied per-request seed did not take effect for any row but the first, so a request's tokens depended on its position in the batch; closes P1 issue #1104, open since 2025-05-31 and labelled `vllm` (C3.4)
**Early stop:** no

### Security
| Q | Answer | Evidence |
|---|--------|----------|
| S1 injection/supply-chain | no | — |
| S2 template overwritten | no | — |
| S3 template obligations | met | all five items ticked; the Tests section states the verification hardware, the counts, and a deliberate negative control |
| S4 agent-directing text | no | — |

### Packaging
| Q | Answer | Evidence |
|---|--------|----------|
| C1.1 dependency bump | no | no supply-bucket changes |
| C1.2 public API changes | no | `flashinfer/sampling.py` is +41/−0 — docstring text plus one added validation. No signature, default or export changes. The *semantics* of existing entry points do change; see the second maintainer note |
| C1.3 AOT/trace registration | no | `aot.py` untouched; no new API and no module generators |

### Presentation
| Q | Answer | Evidence |
|---|--------|----------|
| C2.1 perf claim backed | n-a | no performance claim |

### Implementation
| Q | Answer | Evidence |
|---|--------|----------|
| C3.1 experimental declared | no | — |
| C3.2 durable areas touched | yes — durable | `include/flashinfer/attention/../sampling.cuh` (+82/−45) and `csrc/sampling.cu` (+47/−24) across seven kernels, factored into a shared `ResolvePhiloxSeedOffset`. The edit is mechanical and identical per kernel and reshapes no interface, so this routes `human` rather than `live` |
| C3.3 tests match change | yes | +215 lines in `tests/utils/test_sampling.py` covering six entry points × two vocab sizes plus chain-speculative sampling, broadcast equivalence, reproducibility, and length validation — and, unusually, a negative control: reverting the indexing to `idx = 0` produces 25 failures across all seven kernels, so the tests are demonstrated to catch the bug rather than assumed to |
| C3.4 release-blocker candidate | 🚨 yes | library and test sides both change, which is the strongest blocker signal. Note this is a long-standing contract violation rather than a regression — the kernels never honoured the array, so no release introduced it and no release is exempt |

### Experimental track
| Q | Answer | Evidence |
|---|--------|----------|
| C4.1 declaration obligations | n-a | C3.1 = no |
| C4.2 machine-readable test scope | n-a | C3.1 = no |
| C4.3 isolated from common areas | n-a | C3.1 = no |

**Notes for the maintainer**

- **What is blocking this is an authorization gate, not a technical objection.** GitHub CI is green *on the current head* — the passing run's `head_sha` is `a5998e46`, which is the only commit on the branch, so nothing here is stale. The one thing that has never run is GitLab: the author's trigger was refused on 2026-09-01 with "not authorized to trigger this CI job", and no pipeline has run since. The PR has not moved in three weeks. Someone with permission issuing that run is most of the remaining distance for a P1 that vLLM has been waiting on since May 2025.
- **Behaviour on existing public entry points changes in two ways, neither visible in a signature.** A length-`batch_size` seed tensor was previously accepted and silently ignored beyond row 0; it now takes effect per row, so any caller passing distinct seeds will see different tokens than before — correctly, but differently. Separately, a seed and offset of differing lengths now raise where they previously ran. Both are the right calls; both are worth a line in the release notes rather than discovery by a downstream integrator.
- **Sequence this against #5313.** That PR converts `test_top_k_top_p_sampling_from_probs_logits_alignment` in this exact file to pairwise selection, while this PR adds 215 lines to it — a textual conflict either way. More subtly, this PR reports six `[...128256...]` cases of that same test failing on unmodified `main`; if the pairwise conversion lands first and those cases fall outside the regular selection, a known-failing configuration quietly stops running pre-merge and survives only in the nightly `--full` pass.

<sub>Generated by flashinfer-pr-screen · rubric: docs/code_review_guidance.md · not a code review · AI screening can make mistakes — a maintainer's judgment supersedes this report.</sub>

**Comment by zhaoyangwang-nvidia** [comment]
Thanks @aleozlx for triggering CI! The report shows 23/25 jobs passed, with only the two H100/CUDA 13.4 jobs timing out before producing test results; all 6 multi-GPU/multi-node jobs passed. Could you help retry those two jobs?
